synthtool/languages/php.py

```diff
@@ -0,0 +1,184 @@
+    m = re.search(COPYRIGHT_REGEX, dest)
+    if m:
+        return re.sub(COPYRIGHT_REGEX, f"Copyright {m.group(1)} Google LLC", src, 1)
+    return src
+
+
+def _find_copy_target(src: Path, version_string: str) -> typing.Optional[Path]:
+    """Returns a directory contains the version subdirectory.
+    """
+    logger.debug("_find_copy_target called with %s and %s", src, version_string)
+    entries = os.scandir(src)
+    if not entries:
+        return None
+    for entry in entries:
+        if entry.path.endswith(version_string):
+            return src
+        if entry.is_dir():
+            return _find_copy_target(Path(entry.path).resolve(), version_string)
+    return None
+
+
+def owlbot_copy_version(
+    src: Path, dest: Path, copy_excludes: typing.Optional[typing.List[str]] = None,
+) -> None:
+    """Copies files from a version subdirectory.
+    """
+    logger.debug("owlbot_copy_version called from %s to %s", src, dest)
+
+    if copy_excludes is None:
+        copy_excludes = DEFAULT_COPY_EXCLUDES
+    # detect the version string for later use
+    entries = os.scandir(src / "src")
+    if not entries:
+        logger.info("there is no src directory to copy")
+        return
+    version_string = os.path.basename(os.path.basename(next(entries)))
+    logger.debug("version_string detected: %s", version_string)
+
+    # copy all src including partial veneer classes
+    s.move([src / "src"], dest / "src", merge=_merge, excludes=copy_excludes)
+
+    # copy tests
+    s.move([src / "tests"], dest / "tests", merge=_merge, excludes=copy_excludes)
+
+    # detect the directory containing proto generated PHP source and metadata.
+    entries = os.scandir(src / "proto/src")
+    proto_dir = None
+    metadata_dir = None
+    if not entries:
+        logger.info("there is no proto generated src directory to copy")
+        return
+    for entry in entries:
+        if os.path.basename(entry.path) == METADATA_DIR:
+            metadata_dir = _find_copy_target(Path(entry.path).resolve(), version_string)
+        else:
+            proto_dir = _find_copy_target(Path(entry.path).resolve(), version_string)
+
+    # copy proto files
+    if isinstance(proto_dir, Path):
+        logger.debug("proto_dir detected: %s", proto_dir)
+        s.move([proto_dir], dest / "src", merge=_merge, excludes=copy_excludes)
+
+    # copy metadata files
+    if isinstance(metadata_dir, Path):
+        logger.debug("metadata_dir detected: %s", metadata_dir)
+        s.move([metadata_dir], dest / "metadata", merge=_merge, excludes=copy_excludes)
+
+
+def owlbot_patch() -> None:
+    """Apply some replacements for copied libraries.
+
+    This function assumes the current directory is the target.
+    """
+    logger.debug("owlbot_patch called for %s", os.getcwd())
+
+    # Apply common replacements, currently nothing.
+    pass
+
+
+def owlbot_main(
+    src: Path, dest: Path, copy_excludes: typing.Optional[typing.List[str]] = None,
+) -> None:
```

Comment:
I expect the patch function to be passed as an argument here, defaulting to `owlbot_patch`.